JNDI data sources are sometimes accessed before they are bound when using JPA API with Hibernate as the provider

### Play Version

2.5.10

### API

Java

### Operating System

Windows 8.1

### JDK

java version "1.8.0_40"
Java(TM) SE Runtime Environment (build 1.8.0_40-b26)
Java HotSpot(TM) 64-Bit Server VM (build 25.40-b25, mixed mode)

### Library Dependencies

Hibernate 4.3.11 or 5.2.5

### Expected Behavior

Data sources should be bound to JNDI before DefaultJPAApi / Hibernate attempts to look them up.

### Actual Behavior

Our application uses Play with Java, JPA and Hibernate as the JPA provider. We use injected instances of JPAApi throughout the application to access the JPA API. Intermittently, during application startup, we get an injection exception:
Error injecting constructor, javax.persistence.PersistenceException: Unable to build entity manager factory
...
Caused by: javax.naming.NameNotFoundException: [data source name] not found

where [data source name] is the name of our data source.

Looking at the source code for the JPA implementation in Play I have found the following:

1. play.db.jpa.JPAApi is bound to play.db.jpa.DefaultJPAApi.JPAApiProvider by play.db.jpa.JPAModule
2. play.db.jpa.DefaultJPAApi.JPAApiProvider is a singleton and instantiates play.db.jpa.DefaultJPAApi when it is instantiated itself and calls play.db.jpa.DefaultJPAApi#start() during initialisation.
3. play.db.jpa.DefaultJPAApi#start() immediately attempts to create all persistence units returned by play.db.jpa.JPAConfig#persistenceUnits().
4. javax.persistence.Persistence#createEntityManagerFactory(String) is called for each persistence unit name.
5. As the persistence unit uses a JNDI data source, Hibernate attempts to look up the data source, ultimately calling javax.naming.InitialContext#lookup(javax.naming.InitialContext)

Looking at the connection pool implementation in Play (using the default Hikari connection pool) I have found the following:

1. play.api.db.DBApi is bound to play.api.db.DBApiProvider by play.api.db.DBModule
2. play.api.db.DBApiProvider is a singleton and instantiates play.api.db.DefaultDBApi when it is instantiated itself and calls play.api.db.DefaultDBApi#connect(Boolean) during initialisation.
3. This method attempts to retrieve and immediately close a connection from each configured database, which causes all connection pools to be created by calling the ConnectionPool#create(String, DatabaseConfig, Config) method
4. When the Hikari connection pool is created, it creates the underlying data source and binds it to JNDI by calling javax.naming.InitialContext#rebind(String, Object)

Both lists of actions take place during application startup as Guice is instantiating the bound classes. Here's the injection graph for a Play application (obtained using Guice Grapher):

![guice-graph](https://cloud.githubusercontent.com/assets/366521/21099983/502d04ee-c069-11e6-9880-ab1d329c4aac.png)

From the graph it's clear that the implementations of JPAApi and DBApi do not have any dependencies that would ensure the order in which they are both created, so there is no guarantee that the data source is bound to JNDI by HikariCPConnectionPool before it is accessed by DefaultJPAApi.

### Reproducible Test Case

I have not been able to reproduce this reliably without modifying the Play framework code. However, modifying HikariCPConnectionPool to have JPAApi as a dependency causes the same exception to always appear during startup (as it forces DefaultJPAApi to create the entity manager factory and this in turn causes Hibernate to lookup the data source before it is bound).

--- a/framework/src/play-java-jpa/src/main/java/play/db/jpa/DefaultJPAApi.java
+++ b/framework/src/play-java-jpa/src/main/java/play/db/jpa/DefaultJPAApi.java
@@ -3,6 +3,7 @@
  */
 package play.db.jpa;
 
+import play.db.DBApi;
 import play.inject.ApplicationLifecycle;
 
 import java.util.*;
@@ -36,7 +37,7 @@ public class DefaultJPAApi implements JPAApi {
         private final JPAApi jpaApi;
 
         @Inject
-        public JPAApiProvider(JPAConfig jpaConfig, JPAEntityManagerContext context, ApplicationLifecycle lifecycle) {
+        public JPAApiProvider(JPAConfig jpaConfig, JPAEntityManagerContext context, ApplicationLifecycle lifecycle, DBApi dbApi) {
             // dependency on db api ensures that the databases are initialised
             jpaApi = new DefaultJPAApi(jpaConfig, context);
             lifecycle.addStopHook(() -> {

